datastream-server/src/main/java/com/linkedin/datastream/server/Coordinator.java

```diff
@@ -944,6 +957,101 @@ private void handleLeaderDoAssignment() {
     }
   }
 
+  /**
+   * assign the partition to tasks for a particular datastreamGroup
+   *
+   * @param maybeDatastreamGroupName the datastreamGroup that needs to perform the partition assignment
+   */
+  private void performPartitionAssignment(Optional<String> maybeDatastreamGroupName) {
+    if (!maybeDatastreamGroupName.isPresent()) {
+      _log.error("Datastream group is not found when performing partition assignment");
+      return;
+    }
+
+    boolean succeeded = false;
+    Map<String, Set<DatastreamTask>> previousAssignmentByInstance = new HashMap<>();
+    Map<String, List<DatastreamTask>> newAssignmentsByInstance = new HashMap<>();
+
+    String datastreamGroupName = maybeDatastreamGroupName.get();
+
+    try {
+      previousAssignmentByInstance = _adapter.getAllAssignedDatastreamTasks();
+      Map<String, Set<DatastreamTask>> assignmentByInstance = new HashMap<>(previousAssignmentByInstance);
+
+      // retrieve the datastreamGroups for validatio
+      DatastreamGroup toProcessDatastream =
+          fetchDatastreamGroups().stream().filter(dg -> datastreamGroupName.equals(dg.getName())).findFirst().orElse(null);
+
+      if (toProcessDatastream != null) {
+        AssignmentStrategy strategy = _connectors.get(toProcessDatastream.getConnectorName()).getAssignmentStrategy();
+        Connector connectorInstance = _connectors.get(toProcessDatastream.getConnectorName()).getConnector()
+            .getConnectorInstance();
+        Map<String, Optional<DatastreamGroupPartitionsMetadata>> datastreamPartitions =
+            connectorInstance.getDatastreamPartitions();
+
+        if (datastreamPartitions.containsKey(toProcessDatastream.getName())) {
+          DatastreamGroupPartitionsMetadata subscribes = connectorInstance.getDatastreamPartitions()
+              .get(toProcessDatastream.getName())
+              .orElseThrow(() ->
+                  new DatastreamTransientException("Subscribed partition is not ready yet for datastream " +
+                      toProcessDatastream.getName()));
+
+          assignmentByInstance = strategy.assignPartitions(assignmentByInstance, subscribes);
+        } else {
+          _log.warn("partitions for {} is not found, ignore the partition assignment", toProcessDatastream.getName());
```

Comment:
Should we ignore or retry here? IMO ignoring seems bit dangerous.

Reply:
This case will only happen when the datastream was just paused/removed but we happened to handle the scheduled LEADER_PARTITION_EVENT. In either case we should just ignore and don't retry.

Reply:
Ah thanks for the details, can you add this as a comment in the code, It's really useful.

Reply:
added . thanks.